In [1]:
import pandas as pd

file_path = "C:\\Users\\aoguntuga\\Documents\\Sentiment Analysis Research\\Data\\sentiment_analysis_data.xls"

xl_file = pd.ExcelFile(file_path)
sheetNames = xl_file.sheet_names

In [10]:
dframe1 = xl_file.parse(sheetNames[0])
dframe2 = xl_file.parse(sheetNames[1])
dframe3 = xl_file.parse(sheetNames[2])
dframe1.head()

NoteID  Polarity  Subjectivity  Lab_HbA1C  HbA1C  Compliance Group
0  36493133 -0.072222      0.225000        6.3    6.3                 1
1  37477202 -0.060204      0.097619        6.7    6.7                 1
2  37508754  0.002500      0.082500        6.1    6.1                 1
3  37733788 -0.020833      0.059722        6.5    6.5                 1
4  37789050 -0.008333      0.298016        6.8    6.8                 1

Running tests for normality: Sharpiro-Wilks (if p > 0.05, distribution normally distributed and vice versa if less).

In [11]:
import numpy as np
from scipy import stats

#group 1: Patients with good control over their diabetes
grp1_polarity_array = np.array(dframe1['Polarity'])
grp1_subjectivity_array = np.array(dframe1['Subjectivity'])

#group 2: Patients with fair control over their diabetes
grp2_polarity_array = np.array(dframe2['Polarity'])
grp2_subjectivity_array = np.array(dframe2['Subjectivity'])

#group 3: Patients with poor control over their diabetes
grp3_polarity_array = np.array(dframe3['Polarity'])
grp3_subjectivity_array = np.array(dframe3['Subjectivity'])

In [12]:
#Shapiro-Wilks tests for polarity & subjectivity Group 1
polarity_shapiro_1 = stats.shapiro(grp1_polarity_array)
subjectivity_shapiro_1 = stats.shapiro(grp1_subjectivity_array)

#Shapiro-Wilks tests for polarity & subjectivity Group 2
polarity_shapiro_2 = stats.shapiro(grp2_polarity_array)
subjectivity_shapiro_2 = stats.shapiro(grp2_subjectivity_array)

#Shapiro-Wilks tests for polarity & subjectivity Group 3
polarity_shapiro_3 = stats.shapiro(grp3_polarity_array)
subjectivity_shapiro_3 = stats.shapiro(grp3_subjectivity_array)

In [13]:
print "Grp 1 Polarity normality score (Shapiro-Wilks): %s, p-value: %s" % (polarity_shapiro_1[0],polarity_shapiro_1[1])
print "Grp 1 Subjectivity normality score (Shapiro-Wilks): %s, p-value: %s" % (subjectivity_shapiro_1[0],subjectivity_shapiro_1[1])
print "---------------------------------------------------------------------"
print "Grp 2 Polarity normality score (Shapiro-Wilks): %s, p-value: %s" % (polarity_shapiro_2[0],polarity_shapiro_2[1])
print "Grp 2 Subjectivity normality score (Shapiro-Wilks): %s, p-value: %s" % (subjectivity_shapiro_2[0],subjectivity_shapiro_2[1])
print "---------------------------------------------------------------------"
print "Grp 3 Polarity normality score (Shapiro-Wilks): %s, p-value: %s" % (polarity_shapiro_3[0],polarity_shapiro_3[1])
print "Grp 3 Subjectivity normality score (Shapiro-Wilks): %s, p-value: %s" % (subjectivity_shapiro_3[0],subjectivity_shapiro_3[1])

Grp 1 Polarity normality score (Shapiro-Wilks): 0.876280665398, p-value: 2.59959285001e-12
Grp 1 Subjectivity normality score (Shapiro-Wilks): 0.977149784565, p-value: 0.00135978835169
---------------------------------------------------------------------
Grp 2 Polarity normality score (Shapiro-Wilks): 0.990012526512, p-value: 0.106945179403
Grp 2 Subjectivity normality score (Shapiro-Wilks): 0.960779368877, p-value: 5.04924264533e-06
---------------------------------------------------------------------
Grp 3 Polarity normality score (Shapiro-Wilks): 0.918471574783, p-value: 0.000596269732341
Grp 3 Subjectivity normality score (Shapiro-Wilks): 0.949299871922, p-value: 0.0133532118052


In [23]:
print "Group 1 Polarity > Skewness: %s| Kurtosis: %s" % (stats.describe(grp1_polarity_array).skewness,stats.describe(grp1_polarity_array).kurtosis)
print "Group 1 Subjectivity > Skewness: %s| Kurtosis: %s" % (stats.describe(grp1_subjectivity_array).skewness,stats.describe(grp1_subjectivity_array).kurtosis)
print "--------------------------------------------------------------------------------------------------------------------------------------"
print "Group 2 Polarity > Skewness: %s| Kurtosis: %s" % (stats.describe(grp2_polarity_array).skewness,stats.describe(grp2_polarity_array).kurtosis)
print "Group 2 Subjectivity > Skewness: %s| Kurtosis: %s" % (stats.describe(grp2_subjectivity_array).skewness,stats.describe(grp2_subjectivity_array).kurtosis)
print "-------------------------------------------------------------------------------------------------------------------------------------- "
print "Group 3 Polarity > Skewness: %s| Kurtosis: %s" % (stats.describe(grp3_polarity_array).skewness,stats.describe(grp3_polarity_array).kurtosis)
print "Group 3 Subjectivity > Skewness: %s| Kurtosis: %s" % (stats.describe(grp3_subjectivity_array).skewness,stats.describe(grp3_subjectivity_array).kurtosis)

Group 1 Polarity > Skewness: -1.86270489815| Kurtosis: 14.1387359458
Group 1 Subjectivity > Skewness: -0.0520459897884| Kurtosis: -0.827411678843
--------------------------------------------------------------------------------------------------------------------------------------
Group 2 Polarity > Skewness: -0.0251377378904| Kurtosis: 0.579949406093
Group 2 Subjectivity > Skewness: 0.101575597076| Kurtosis: -1.06098828334
-------------------------------------------------------------------------------------------------------------------------------------- 
Group 3 Polarity > Skewness: 0.729709247063| Kurtosis: 4.53355485772
Group 3 Subjectivity > Skewness: -0.397242897675| Kurtosis: -0.523869122848


Polarity Skewness: -1.86, Polarity Kurtosis: 14.14
Subjectivity Skewness: -0.05, Subjectivity Kurtosis: -0.82
For a distribution to be normal Skewness needs to be btw -0.5 & 0.5 and Kurtosis  should be equals to 3
Polarity and Subjectivity distributions for group 1 (People with good control) is not normally distributed. Pearson correlation cannot be used to test for correlation btwn HbA1C and each value. The Spearman correlation is a better test.

In [7]:
grp1_hba1c_array = np.array(dframe1['Lab_HbA1C'])

#Creating ranked distributions for HbA1C, Polarity and Subjectivity for Group 1
grp_1_polar_data_ranked = stats.stats.rankdata(grp1_polarity_array)
grp_1_subjectibity_data_ranked = stats.stats.rankdata(grp1_subjectivity_array)
grp_1_hba1c_data_ranked = stats.stats.rankdata(grp1_hba1c_array)

#Spearman Rho calculation using ranked distributions for Group 1
grp1_correlation_polarity = stats.spearmanr(grp_1_hba1c_data_ranked,grp_1_polar_data_ranked)
grp1_correlation_subjectivity = stats.spearmanr(grp_1_hba1c_data_ranked,grp_1_subjectibity_data_ranked)

In [8]:
print "Group 1 HbA1c & Polarity correlation (Spearman Rho): %s, p-value: %s" % (grp1_correlation_polarity[0],grp1_correlation_polarity[1])
print "Group 1 HbA1c & Subjectivity correlation (Spearman Rho): %s, p-value: %s" % (grp1_correlation_subjectivity[0],grp1_correlation_subjectivity[1])

HbA1c & Polarity correlation (Spearman Rho): -0.0596442140184, p-value: 0.381944983317
HbA1c & Subjectivity correlation (Spearman Rho): -0.0201430718391, p-value: 0.767961705302


In [24]:
grp2_hba1c_array = np.array(dframe2['Lab_HbA1C'])

#Creating ranked distributions for HbA1C, Polarity and Subjectivity for Group 2
grp_2_polar_data_ranked = stats.stats.rankdata(grp2_polarity_array)
grp_2_subjectibity_data_ranked = stats.stats.rankdata(grp2_subjectivity_array)
grp_2_hba1c_data_ranked = stats.stats.rankdata(grp2_hba1c_array)

#Spearman Rho calculation using ranked distributions Group 2.
grp2_correlation_polarity = stats.spearmanr(grp_2_hba1c_data_ranked,grp_2_polar_data_ranked)
grp2_correlation_subjectivity = stats.spearmanr(grp_2_hba1c_data_ranked,grp_2_subjectibity_data_ranked)

print "Group 2 HbA1c & Polarity correlation (Spearman Rho): %s, p-value: %s" % (grp2_correlation_polarity[0],grp2_correlation_polarity[1])
print "Group 2 HbA1c & Subjectivity correlation (Spearman Rho): %s, p-value: %s" % (grp2_correlation_subjectivity[0],grp2_correlation_subjectivity[1])

Group 2 HbA1c & Polarity correlation (Spearman Rho): 0.0794526403135, p-value: 0.225972729912
Group 2 HbA1c & Subjectivity correlation (Spearman Rho): 0.117585650901, p-value: 0.0726034489114


In [25]:
grp3_hba1c_array = np.array(dframe3['Lab_HbA1C'])

#Creating ranked distributions for HbA1C, Polarity and Subjectivity for Group 3
grp_3_polar_data_ranked = stats.stats.rankdata(grp3_polarity_array)
grp_3_subjectibity_data_ranked = stats.stats.rankdata(grp3_subjectivity_array)
grp_3_hba1c_data_ranked = stats.stats.rankdata(grp3_hba1c_array)

#Spearman Rho calculation using ranked distributions Group 3.
grp3_correlation_polarity = stats.spearmanr(grp_3_hba1c_data_ranked,grp_3_polar_data_ranked)
grp3_correlation_subjectivity = stats.spearmanr(grp_3_hba1c_data_ranked,grp_3_subjectibity_data_ranked)

print "Group 3 HbA1c & Polarity correlation (Spearman Rho): %s, p-value: %s" % (grp3_correlation_polarity[0],grp3_correlation_polarity[1])
print "Group 3 HbA1c & Subjectivity correlation (Spearman Rho): %s, p-value: %s" % (grp3_correlation_subjectivity[0],grp3_correlation_subjectivity[1])

Group 3 HbA1c & Polarity correlation (Spearman Rho): -0.00838189514311, p-value: 0.948881469981
Group 3 HbA1c & Subjectivity correlation (Spearman Rho): 0.0435502751483, p-value: 0.738938055861
